### PR TITLE
GODRIVER-2911: Add machine flow OIDC authentication

### PR DESCRIPTION
**Pull request by pmeredit**
GODRIVER-2911

## Summary

This is a full implementation of machine flow OIDC with token invalidation and reauthentication. It is missing configuration enforcement, which is split into GODRIVER-3249 to keep the PR smaller.

This currently does not have the tests, but I wanted to give an opportunity for architecture feedback. This should work since it's following the same algorithm as the rust code, but there could be something I'm missing here with, for instance, how reauthentication should work since this is different than how the rust driver does auth, and I want to make sure this looks correct.

## Biggest changes
* Several types moved into driver package
* Authenticators added to all `Command`s and `Operation`s to handle reauth
* Two methods added to `Connection` interface (:()

## Background & Motivation
epic: https://jira.mongodb.org/browse/GODRIVER-2574

### CODE REVIEW

File: x/mongo/driver/auth/oidc.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+const environmentProp = "ENVIRONMENT"
+const allowedHostsProp = "ALLOWED_HOSTS"
+
+const azureEnvironmentValue = "azure"
+const gcpEnvironmentValue = "gcp"
+
+const apiVersion = 1
+const invalidateSleepTimeout = 100 * time.Millisecond
+const machineCallbackTimeout = 60 * time.Second
+
```
Comment (pmeredit) [inline x/mongo/driver/auth/oidc.go:24]: I just ported over all the constants, some of these will not be used until other tickets like Human flow

File: x/mongo/driver/driver.go [hunk 2]
```diff
@@ -24,6 +25,62 @@ import (
...
+}
+
+type IDPInfo struct {
+	Issuer        string   `bson:"issuer"`
+	ClientID      string   `bson:"clientId"`
+	RequestScopes []string `bson:"requestScopes"`
+}
+
+// Authenticator handles authenticating a connection. The implementors of this interface
+// are all in the auth package. Most authentication mechanisms to not allow for Reauth,
```
Comment (blink1073) [inline x/mongo/driver/driver.go:66]: mechanisms do not allow*

**Review by blink1073 — COMMENTED** [review]
The approach looks good to me so far!

**Comment by blink1073** [comment]
For linting, see https://github.com/mongodb/mongo-go-driver/blob/v1/docs/CONTRIBUTING.md#linting-on-commit

File: x/mongo/driver/driver.go [hunk 3]
```diff
@@ -24,6 +25,65 @@ import (
...
+	RefreshToken *string
+}
+
+// IDPInfo contains the information needed to perform OIDC authentication with an Identity Provider.
+type IDPInfo struct {
+	Issuer        string   `bson:"issuer"`
+	ClientID      string   `bson:"clientId"`
+	RequestScopes []string `bson:"requestScopes"`
+}
+
```
Comment (pmeredit) [inline x/mongo/driver/driver.go:65]: I think this is my biggest question on design right here. I could add another interface called `Reauthenticator` and downcast in operation.go on a 391 error to `Reauthenticator` so that every auth mechanism doesn't need to have a definition for `Reauth`, but I like this approach since it will require any new authentication mechanism to consider whether or not it supports Reauth than simply forgetting to do that, since this moves the check to compile time instead of runtime. This is the rust programmer in me, so I'll defer to whatever others prefer

File: etc/run-oidc-test.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
```
Comment (pmeredit) [inline etc/run-oidc-test.sh:1]: I made this an executable rather than using `go test` since this is what gssapi is doing... but also I know that when I go to test GCP and Azure that I'm going to want to build ahead of time, and I'm not sure if `go test` can build an executable (`cargo test` can).

File: cmd/testoidcauth/main.go [hunk 5]
```diff
@@ -0,0 +1,676 @@
+// Copyright (C) MongoDB, Inc. 2022-present.
```
Comment (pmeredit) [inline cmd/testoidcauth/main.go:1]: I made this an executable rather than using go test since this is what gssapi is doing... but also I know that when I go to test GCP and Azure that I'm going to want to build ahead of time, and I'm not sure if go test can build an executable (cargo test can).

Comment (pmeredit) [reply cmd/testoidcauth/main.go:1]: Also, here's hoping statically linking glibc works with go, since that was necessary to get the rust tests working :)

File: x/mongo/driver/driver.go (hunk 3 above)
Comment (blink1073) [reply x/mongo/driver/driver.go:65]: I agree, and the server may also decide to add Reauth capability to existing mechanisms.

File: cmd/testoidcauth/main.go (hunk 5 above)
Comment (matthewdale) [reply cmd/testoidcauth/main.go:1]: `go test` can indeed build a binary. However, following the current convention seems fine. If/when we want to make these Go tests, we can convert them all at the same time.

File: x/mongo/driver/driver.go (hunk 3 above)
Comment (matthewdale) [reply x/mongo/driver/driver.go:65]: I most often see "upgrading" Go interfaces used as a backward compatibility mechanism. IMO some extra no-op functions is better than runtime type assertions. I agree with the current approach.

File: x/mongo/driver/auth/oidc.go [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+func (oa *OIDCAuthenticator) doAuthHuman(_ context.Context, _ *Config, _ OIDCCallback) error {
+	// TODO GODRIVER-3246: Implement OIDC human flow
+	// Println is for linter
+	fmt.Println("OIDC human flow not implemented yet", oa.idpInfo)
+	return newAuthError("OIDC human flow not implemented yet", nil)
+}
+
+func (oa *OIDCAuthenticator) doAuthMachine(ctx context.Context, cfg *Config, machineCallback OIDCCallback) error {
+	accessToken, err := oa.getAccessToken(ctx,
+		&OIDCArgs{Version: apiVersion,
```
Comment (matthewdale) [inline x/mongo/driver/auth/oidc.go:303]: Optional: Format for easier reading.

```suggestion
		&OIDCArgs{
			Version: apiVersion,
```

File: x/mongo/driver/auth/oidc.go [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+	defer oa.mu.Unlock()
+	var err error
+
+	if cfg == nil {
+		return newAuthError(fmt.Sprintf("config must be set for %q authentication", MongoDBOIDC), nil)
+	}
+	oa.cfg = cfg
+
+	if oa.accessToken != "" {
+		err = ConductSaslConversation(ctx, cfg, "$external", &oidcOneStep{
```
Comment (matthewdale) [inline x/mongo/driver/auth/oidc.go:262]: Holding an exclusive lock while we run the SASL conversation will limit the auth part of connection establishment to one-at-a-time, overriding the intended behavior of `maxConnecting`. Instead, we should only hold a lock when invalidating and acquiring a new access token.

Unfortunately that means the strategy of storing the `auth.Config` on the `OIDCAuthenticator` also won't work.

E.g. changes to prevent locking during the SASL conversation and allow multiple `Config` instances:
```go
func (oa *OIDCAuthenticator) invalidateAccessToken(tokenGenID uint64) {
	oa.mu.Lock()
	defer oa.mu.Unlock()

	if oa.tokenGenID <= tokenGenID {
		oa.accessToken = ""
	}
	// ...
}

func (oa *OIDCAuthenticator) getAccessToken(...) (string, uint64, error) {
	oa.mu.Lock()
	defer oa.mu.Unlock()

	if oa.accessToken != "" {
		return oa.accessToken, oa.tokenGenID, nil
	}
	// ...
}

func (oa *OIDCAuthenticator) Auth(ctx context.Context, cfg *Config) error {
	var err error

	if cfg == nil {
		return newAuthError(fmt.Sprintf("config must be set for %q authentication", MongoDBOIDC), nil)
	}

	accessToken, tokenGenID, err := oa.getAccessToken(...)

	if oa.accessToken != "" {
		err = ConductSaslConversation(ctx, cfg, "$external", &oidcOneStep{
			userName:    oa.userName,
			accessToken: oa.accessToken,
		})
		if err == nil {
			return nil
		}
		oa.invalidateAccessToken(tokenGenID)
		time.Sleep(invalidateSleepTimeout)
	}

	oa.mu.Lock()
	defer oa.mu.Unlock()
	// ...
}
```

File: x/mongo/driver/auth/oidc.go [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+	tokenGenID := oa.cfg.Connection.OIDCTokenGenID()
+	if tokenGenID >= oa.tokenGenID {
+		oa.accessToken = ""
+		oa.cfg.Connection.SetOIDCTokenGenID(0)
+	}
+}
+
+// Reauth reauthenticates the connection when the server returns a 391 code. Reauth is part of the
+// driver.Authenticator interface.
+func (oa *OIDCAuthenticator) Reauth(ctx context.Context) error {
```
Comment (matthewdale) [inline x/mongo/driver/auth/oidc.go:237]: `Reauth` needs to take accept an `AuthConfig` that has all the inputs for authenticating a specific connection, not the connection most recently configured in `oa.cfg`.

E.g.
```go
func (oa *OIDCAuthenticator) Reauth(
	ctx context.Context,
	cfg *AuthConfig,
) error {
```

It's not clear exactly how to get all the info for an `AuthConfig` in `Operation.Execute`.

File: cmd/testoidcauth/main.go [hunk 9]
```diff
@@ -0,0 +1,676 @@
...
+	})
+
+	defer client.Disconnect(context.Background())
+
+	if err != nil {
+		return fmt.Errorf("machine_3_1: failed connecting client: %v", err)
+	}
+
+	// Poison the cache with a random token
+	client.GetAuthenticator().(*auth.OIDCAuthenticator).SetAccessToken("some random happy sunshine string")
```
[the remark this answers is not in the document]
Comment (pmeredit) [reply cmd/testoidcauth/main.go:325]: I didn't know you could do this, thanks, much preferable

File: x/mongo/driver/operation/insert.go [hunk 10]
```diff
@@ -47,6 +48,7 @@ type Insert struct {
 
 // InsertResult represents an insert result returned by the server.
 type InsertResult struct {
+	authenticator driver.Authenticator
```
[the remark this answers is not in the document]
Comment (pmeredit) [reply x/mongo/driver/operation/insert.go:51]: I added these with a python script and missed this one :)

File: x/mongo/driver/driver.go [hunk 11]
```diff
@@ -24,6 +25,65 @@ import (
...
+// OIDCCallback is the type for both Human and Machine Callback flows. RefreshToken will always be
+// nil in the OIDCArgs for the Machine flow.
+type OIDCCallback func(context.Context, *OIDCArgs) (*OIDCCredential, error)
+
+// OIDCArgs contains the arguments for the OIDC callback.
+type OIDCArgs struct {
+	Version      int
+	Timeout      time.Time
```
[the remark this answers is not in the document]
Comment (pmeredit) [reply x/mongo/driver/driver.go:47]: Nope, I think you're right

File: x/mongo/driver/auth/oidc.go (hunk 8 above)
Comment (pmeredit) [reply x/mongo/driver/auth/oidc.go:237]: I don't see how we can make that work. The python and rust driver both store the necessary information in the Authenticator :/

File: x/mongo/driver/auth/oidc.go (hunk 7 above)
Comment (pmeredit) [reply x/mongo/driver/auth/oidc.go:262]: I think this works, let me run the tests. My concern is machine_1_2, but I do think this works.

File: x/mongo/driver/operation/insert.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (pmeredit) [reply x/mongo/driver/operation/insert.go]: Looks like I added a few other extra ones, should all be gone now

File: x/mongo/driver/auth/oidc.go (hunk 7 above)
Comment (pmeredit) [reply x/mongo/driver/auth/oidc.go]: This worked

File: Makefile [hunk 12]
```diff
@@ -132,6 +132,10 @@ evg-test-atlas-data-lake:
 evg-test-enterprise-auth:
 	go run -tags gssapi ./cmd/testentauth/main.go
 
+.PHONY: evg-test-oidc-auth
+evg-test-oidc-auth:
+	go run ./cmd/testoidcauth/main.go
```
Comment (matthewdale) [inline Makefile:137]: Since we're running this not as part of a standard test suite, we need to explicitly add a 2nd run with the data race detector enabled.

```suggestion
	go run ./cmd/testoidcauth/main.go
	go run -race ./cmd/testoidcauth/main.go
```

P.S. Yes, we should also do this for all of the other `go run ...` make targets, but we can do that in another PR.

File: x/mongo/driver/auth/oidc.go [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+}
+
+// Auth authenticates the connection.
+func (oa *OIDCAuthenticator) Auth(ctx context.Context, cfg *Config) error {
+	var err error
+
+	if cfg == nil {
+		return newAuthError(fmt.Sprintf("config must be set for %q authentication", MongoDBOIDC), nil)
+	}
+	oa.httpClient = cfg.HTTPClient
```
Comment (matthewdale) [inline x/mongo/driver/auth/oidc.go:264]: There's a data race detected here when running the OIDC tests. If possible, we should pass in the HTTP client when we create the authenticator at initialization rather than when we call `Auth` to avoid having to write to this field at run time.

See [data-race-1.txt](https://github.com/user-attachments/files/16057070/data-race-1.txt) for details.

File: x/mongo/driver/auth/oidc.go [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+	if oa.accessToken != "" {
+		return oa.accessToken, nil
+	}
+
+	cred, err := callback(ctx, args)
+	if err != nil {
+		return "", err
+	}
+
+	oa.accessToken = cred.AccessToken
```
Comment (matthewdale) [inline x/mongo/driver/auth/oidc.go:193]: There's a data race detected here when running the OIDC tests. We either need to use an `atomic.Value` to synchronize access to the access token value, or delegate all reading and writing of the access token to a function that synchronizes access to it.

See [data-race-2.txt](https://github.com/user-attachments/files/16057090/data-race-2.txt) for details.

**Review by matthewdale — CHANGES_REQUESTED** [review]

File: x/mongo/driver/auth/oidc.go [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+		},
+		machineCallback)
+	cancel()
+	if err != nil {
+		return err
+	}
+	return runSaslConversation(ctx,
+		cfg,
+		newSaslConversation(&oidcOneStep{accessToken: accessToken}, "$external", false),
+	)
```
Comment (matthewdale) [inline x/mongo/driver/auth/oidc.go:325]: Can this use `ConductSaslConversation` instead? That would allow merging `runSaslConversation` and `ConductSaslConversation` together.

E.g.
```go
return ConductSaslConversation(
	ctx,
	cfg,
	"$external",
	&oidcOneStep{accessToken: accessToken})
```

File: x/mongo/driver/auth/oidc.go [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+	if machineCallback != nil {
+		return oa.doAuthMachine(ctx, cfg, machineCallback)
+	}
+	return newAuthError("no OIDC callback provided", nil)
+}
+
+func (oa *OIDCAuthenticator) doAuthHuman(_ context.Context, _ *Config, _ OIDCCallback) error {
+	// TODO GODRIVER-3246: Implement OIDC human flow
+	// Println is for linter
+	fmt.Println("OIDC human flow not implemented yet", oa.idpInfo)
```
Comment (matthewdale) [inline x/mongo/driver/auth/oidc.go:303]: Remove print statement.

```suggestion
```

File: x/mongo/driver/auth/oidc.go [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+	return newAuthError("OIDC human flow not implemented yet", nil)
+}
+
+func (oa *OIDCAuthenticator) doAuthMachine(ctx context.Context, cfg *Config, machineCallback OIDCCallback) error {
+	subCtx, cancel := context.WithTimeout(ctx, machineCallbackTimeout)
+	accessToken, err := oa.getAccessToken(subCtx,
```
Comment (matthewdale) [inline x/mongo/driver/auth/oidc.go:309]: The line in the OIDC spec that this timeout comes from is ambiguous with respect to the Go Driver:
> If CSOT is not applied, then the driver MUST use 1 minute as the timeout.

All Go Driver blocking APIs support timeouts, so from that perspective, CSOT is always applied. However, the Go Driver also includes the concept of CSOT being "enabled" or not, which was required to retrofit the CSOT behavior into the existing driver that already supported timeouts (for example, see usage of `csot.IsTimeoutContext`). From that perspective, we should only apply the 1-minute timeout when CSOT is not "enabled".

As implemented, the code always applies the 1-minute timeout, overriding any longer timeouts. That seems unlikely to break anything because waiting 1 minute for an access token will probably cause other problems (e.g. the default `connectTimeout` is 30 seconds).

Overall, I think this code is safe, but is worth a comment describing the interpretation of the OIDC spec we're using here.

E.g. comment:
> The CSOT spec says to apply a 1-minute timeout if "CSOT is not applied". That's ambiguous for the v1.x Go Driver because it could mean either "no timeout provided" or "CSOT not enabled". Always use a maximum timeout duration of 1 minute, allowing us to ignore the ambiguity. Contexts with a shorter timeout are unaffected.

**Review by matthewdale — COMMENTED** [review]
Some small improvement suggestions, but the overall behavior looks good!

File: x/mongo/driver/auth/oidc.go (hunk 16 above)
Comment (pmeredit) [reply x/mongo/driver/auth/oidc.go:303]: The linter fails if I remove the print ;)

File: x/mongo/driver/auth/oidc.go (hunk 17 above)
Comment (pmeredit) [reply x/mongo/driver/auth/oidc.go:309]: This is exactly the sort of context I am missing, thank you!

File: mongo/options/clientoptions.go [hunk 18]
```diff
@@ -110,6 +110,8 @@ type Credential struct {
 	Username                string
 	Password                string
 	PasswordSet             bool
+	OIDCMachineCallback     driver.OIDCCallback
+	OIDCHumanCallback       driver.OIDCCallback
```
Comment (matthewdale) [inline mongo/options/clientoptions.go:114]: We need to canonically define or re-define the `OIDCCallback` type and associated structs here so the public API doesn't include symbols from the experimental API.

E.g. types that need to be defined in the `options` package:
```go
// OIDCCallback is the type for both Human and Machine Callback flows. RefreshToken will always be
// nil in the OIDCArgs for the Machine flow.
type OIDCCallback func(context.Context, *OIDCArgs) (*OIDCCredential, error)

// OIDCArgs contains the arguments for the OIDC callback.
type OIDCArgs struct {
	Version      int
	IDPInfo      *IDPInfo
	RefreshToken *string
}

// OIDCCredential contains the access token and refresh token.
type OIDCCredential struct {
	AccessToken  string
	ExpiresAt    *time.Time
	RefreshToken *string
}
```
It may be possible to canonically define the `OIDCCallback`, `OIDCArgs`, and `OIDCCredential` types in the `options` package, and use those types in the `driver` and auth` packages. If not, we will have to add logic to convert between the types.

**Review by matthewdale — CHANGES_REQUESTED** [review]

File: mongo/options/clientoptions.go (hunk 18 above)
Comment (pmeredit) [reply mongo/options/clientoptions.go]: Tell me if this solution works for you, I think it's good.

**Review by matthewdale — APPROVED** [review]
Looks good! 👍
